### PR TITLE
ChrPlasmaLens element: Support k<0, and fix uninitialized-variable bug in tracking of reference particle

### PR DESCRIPTION
**Pull request by kyrsjo**
This patch adds support for negative strength (defocusing) for `ChrPlasmaLens` and `ConstF` elements. It also fixes a bug (use of uninitialized variable) in reference particle tracking for `ChrPlasmaLens`.

### CODE REVIEW

**Comment by kyrsjo** [comment]
Hm, as soon as I made the PR I realized I forgot to fix tabs / spaces. However the bot seems to have messed up the number of spaces needed.

I'll fix (and probably force push to override its broken fixes).

Looks like the CI is running the tests for me, that's nice :)

**Comment by ax3l** [comment]
Thank you for your first PR to ImpactX, @kyrsjo!

Those changes look great, let us structure & test them to make them best documented/portable/covered by CI.
Please simply push additional commits as needed to the PR to update it as needed.

I also started CI and [sent you an invite to our GitHub org](https://github.com/BLAST-ImpactX), please accept to auto-run CI even before your first PR got merged on updates.

Please make sure to provide a [detailed description on the change](https://github.com/BLAST-ImpactX/impactx/pull/1030#issue-3201891770) and try to do one change at a time. e.g., if you fix a bug do it in one PR, adding a feature in another. Ideally, if it affects multiple elements as in your case, with unrelated bugs, then please make a PR per element (plasma lens, quad and ConstF here could be each a PR.)
This makes it easier for us to document changes for users and we can also "backport" fixes to older releases, when needed. It also enables us to merge things faster, because they are less complex and orthogonal changes can progress in parallel.

Please feel free to add your name to the author list of the respective files you changed.

If a specific bug is fixed (or feature added), then I usually structure my PR (branch) like this:
- a commit that adds a test that triggers the exact bug (to cover it) / covers the new feature
- another commit that adds the fix / feature
- documentation updates, if needed (usually for new features)

Welcome again, looking forward to merging those! Please let us know if you have any questions :sparkles:

File: src/elements/ChrPlasmaLens.H [hunk 1]
```diff
@@ -152,21 +152,43 @@ namespace impactx::elements
...
-            auto const [sin_ods, cos_ods] = amrex::Math::sincos(omega*m_slice_ds);
-
-            // advance transverse position and momentum (focusing)
-            xout = cos_ods * x + sin_ods / (omega * delta1) * px;
-            pxout = -omega * delta1 * sin_ods * x + cos_ods * px;
-
-            yout = cos_ods * y + sin_ods / (omega * delta1) * py;
-            pyout = -omega * delta1 * sin_ods * y + cos_ods * py;
-
+            if (m_g > 0)
```
Comment (ax3l) [inline src/elements/ChrPlasmaLens.H]: Just to use the right literal type:
```suggestion
            if (m_g > 0_prt)
```

File: src/elements/ChrPlasmaLens.H [hunk 2]
```diff
@@ -152,21 +152,43 @@ namespace impactx::elements
...
+
+                // advance transverse position and momentum (focusing)
+                xout = cos_ods * x + sin_ods / (omega * delta1) * px;
+                pxout = -omega * delta1 * sin_ods * x + cos_ods * px;
+
+                yout = cos_ods * y + sin_ods / (omega * delta1) * py;
+                pyout = -omega * delta1 * sin_ods * y + cos_ods * py;
+
+            }
+            else if (m_g < 0)
```
Comment (ax3l) [inline src/elements/ChrPlasmaLens.H]:
```suggestion
            else if (m_g < 0_prt)
```

File: src/elements/ConstF.H [hunk 3]
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
...
             // trigo
-            auto const [sin_kxds, cos_kxds] = amrex::Math::sincos(m_kx * slice_ds);
-            m_cos_kxds = cos_kxds;
-            m_const_x = -m_kx * sin_kxds;
-            auto const [sin_kyds, cos_kyds] = amrex::Math::sincos(m_ky * slice_ds);
-            m_cos_kyds = cos_kyds;
-            m_const_y = -m_ky * sin_kyds;
+            // Note: The convention here is m_kx[1/m] = sign(k)*sqrt(abs(k))
+            // where k[1/m^2] is in the usual MADX convention
+            if (m_kx > 0)
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            if (m_kx > 0_prt)
```

File: src/elements/ConstF.H [hunk 4]
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
...
+            {
+              auto const [sin_kxds, cos_kxds] = amrex::Math::sincos(m_kx * slice_ds);
+              m_cos_kxds = cos_kxds;
+              m_const_x  = -m_kx * sin_kxds;
+              m_sincx    = sin_kxds / m_kx;
+            }
+            else if (m_kx < 0)
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            else if (m_kx < 0_prt)
```

File: src/elements/ConstF.H [hunk 5]
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
...
+              m_sincx    = -sinh_kxds / m_kx;
+            }
+            else //m_kx == 0
+            {
+              m_cos_kxds = 1.0;
+              m_const_x  = 0.0;
+              m_sincx    = slice_ds;
+            }
+
+            if (m_ky > 0)
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            if (m_ky > 0_prt)
```

File: src/elements/ConstF.H [hunk 6]
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
...
+            {
+              auto const [sin_kyds, cos_kyds] = amrex::Math::sincos(m_ky * slice_ds);
+              m_cos_kyds = cos_kyds;
+              m_const_y  = -m_ky * sin_kyds;
+              m_sincy    = sin_kyds / m_ky;
+            }
+            else if (m_ky < 0)
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            else if (m_ky < 0_prt)
```

File: src/elements/ConstF.H [hunk 7]
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
...
+              m_sincy    = -sinh_kyds / m_ky;
+            }
+            else //m_ky == 0
+            {
+              m_cos_kyds = 1.0;
+              m_const_y  = 0.0;
+              m_sincy    = slice_ds;
+            }
+
+            if (m_kt < 0)
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            if (m_kt < 0_prt)
```

File: src/elements/ConstF.H [hunk 8]
```diff
@@ -99,19 +99,60 @@ namespace impactx::elements
...
+            {
+              throw std::runtime_error(std::string(type) + ": must have kt >= 0!");
```
Comment (ax3l) [inline src/elements/ConstF.H]: We do not automatically enforce it, but just FYI: We use four spaces for indentation :)

File: src/elements/ConstF.H [hunk 9]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
             // trigo
-            auto const [sin_kxds, cos_kxds] = amrex::Math::sincos(m_kx * slice_ds);
-            amrex::ParticleReal const_x = -m_kx * sin_kxds;
-            auto const [sin_kyds, cos_kyds] = amrex::Math::sincos(m_ky * slice_ds);
-            amrex::ParticleReal const_y = -m_ky * sin_kyds;
+            // Note: The convention here is m_kx[1/m] = sign(k)*sqrt(abs(k))
+            // where k[1/m^2] is in the usual MADX convention
+
+            amrex::ParticleReal sin_kxds, cos_kxds, const_x, sincx;
+            if (m_kx > 0) {
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            if (m_kx > 0_prt) {
```

File: src/elements/ConstF.H [hunk 10]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
+              std::tie(sin_kxds, cos_kxds) = amrex::Math::sincos(m_kx * slice_ds);
+              const_x = -m_kx * sin_kxds;
+              sincx   = sin_kxds / m_kx;
+            }
+            else if (m_kx < 0) {
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            else if (m_kx < 0_prt) {
```

File: src/elements/ConstF.H [hunk 11]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
+              sin_kxds = std::sinh(-m_kx*slice_ds);
+              cos_kxds = std::cosh(-m_kx*slice_ds);
+              const_x  = -m_kx * sin_kxds;
+              sincx    = -sin_kxds / m_kx;
+            }
+            else { //m_kx == 0
+              cos_kxds = 1.0;
+              const_x  = 0.0;
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
              cos_kxds = 1_prt;
              const_x  = 0_prt;
```

File: src/elements/ConstF.H [hunk 12]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
+              sincx    = slice_ds;
+            }
+
+            amrex::ParticleReal sin_kyds, cos_kyds, const_y, sincy;
+            if (m_ky > 0) {
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            if (m_ky > 0_prt) {
```

File: src/elements/ConstF.H [hunk 13]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
+              std::tie(sin_kyds, cos_kyds) = amrex::Math::sincos(m_ky * slice_ds);
+              const_y = -m_ky * sin_kyds;
+              sincy   = sin_kyds / m_ky;
+            }
+            else if (m_ky < 0) {
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            else if (m_ky < 0_prt) {
```

File: src/elements/ConstF.H [hunk 14]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
+              sin_kyds = std::sinh(-m_ky*slice_ds);
+              cos_kyds = std::cosh(-m_ky*slice_ds);
+              const_y  = -m_ky * sin_kyds;
+              sincy    = -sin_kyds / m_ky;
+            }
+            else { //m_ky == 0
+              cos_kyds = 1.0;
+              const_y  = 0.0;
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
              cos_kyds = 1_prt;
              const_y  = 0_prt;
```

File: src/elements/ConstF.H [hunk 15]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
+              sincy    = slice_ds;
+            }
+
+            if (m_kt < 0) {
```
Comment (ax3l) [inline src/elements/ConstF.H]:
```suggestion
            if (m_kt < 0_prt) {
```

File: src/elements/ConstF.H [hunk 16]
```diff
@@ -237,29 +278,67 @@ namespace impactx::elements
...
-
             // intermediate quantities - to avoid division by zero
-            amrex::ParticleReal sincx = m_kx > 0 ? std::sin(m_kx * slice_ds) / m_kx : slice_ds;
-            amrex::ParticleReal sincy = m_ky > 0 ? std::sin(m_ky * slice_ds) / m_ky : slice_ds;
             amrex::ParticleReal const sinct = m_kt > 0 ? std::sin(m_kt * slice_ds) / m_kt : slice_ds;
             amrex::ParticleReal const_pt = sinct / betgam2;
 
             // assign linear map matrix elements
             Map6x6 R = Map6x6::Identity();
+
```
Comment (ax3l) [inline src/elements/ConstF.H]: Not really needed change, but ok to leave in, looks sensible.

File: src/elements/Quad.H [hunk 17]
```diff
@@ -271,9 +271,9 @@ namespace impactx::elements
                 R(4,4) = std::cos(omega*slice_ds);
                 R(5,6) = slice_ds/betgam2;
             } else {
-                R(1,2) = m_slice_ds;
-                R(3,4) = m_slice_ds;
-                R(5,6) = m_slice_ds / betgam2;
+                R(1,2) = slice_ds;
```
Comment (ax3l) [inline src/elements/Quad.H:284]: Quad transport map fix for slicing: Please in a separate PR.

File: src/elements/ChrPlasmaLens.H [hunk 18]
```diff
@@ -152,21 +152,43 @@ namespace impactx::elements
...
+            {
+                auto const sinh_ods = std::sinh(omega*m_slice_ds);
+                auto const cosh_ods = std::cosh(omega*m_slice_ds);
+
+                // advance transverse position and momentum (defocusing)
+                xout = cosh_ods * x + sinh_ods / (omega * delta1) * px;
```
Comment (ax3l) [inline src/elements/ChrPlasmaLens.H:205]: The logic for the `R(1,1)`, `R(1,2)` etc branches here looks very similar for positive and negative values of `m_g`.

It might be worth assigning the intermediate values to variables `R_11`, `R_12`, etc., and write this branch-less (via ternary operators)/with shorter branches to:
- avoid code duplication
- reduce register pressure on GPUs

File: src/elements/ConstF.H (hunk 8 above)
Comment (kyrsjo) [reply src/elements/ConstF.H]: Oops. Missed one!

File: src/elements/ChrPlasmaLens.H (hunk 18 above)
Comment (kyrsjo) [reply src/elements/ChrPlasmaLens.H:205]: Sure, I was assuming that the compiler would smooth it out anyway.

And putting them in blocks like this meant I could use both const, and the `amrex::sincos` combined call (while sinh/cosh needs two calls). For the actual code, the position formulas are essentially identical, but the px has a different sign.

So it will be a lot of branches, and anyway the k=0 case is different. So if you insist I will rewrite with ternaries, but I do think it's better like this.

File: src/elements/ConstF.H (hunk 8 above)
Comment (kyrsjo) [reply src/elements/ConstF.H]: Your tabs-to-spaces autoconverter did something else though, I think!

( I realized I hadn't fixed my editor config after reinstalling the laptop when your CI yelled at me and grumpily tried to to fixed it :P)

**Comment by kyrsjo** [comment]
Hi @ax3l , thanks for the invite and for looking over it! Most fixes are easy, will do. For the "use ternary operator" thing, please see my comments.

I guess I should add tests for negative-k cases of these elements also.

Is it OK that we keep the elements in 1 PR? I'll separate out the `m_slice_ds` uninitialized stuff to a separate PR.

I see that the `apochromat_pl` and `apochromat_pl.py` tests are failing, will look into that.

General comment: I appreciate that you are insisting on small and readable PRs! Having been involved in a few codes myself, especially being the former release manager and git-repo-czar of sixtrack, it's a very good thing.

File: src/elements/ChrPlasmaLens.H (hunk 18 above)
Comment (cemitch99) [reply src/elements/ChrPlasmaLens.H:205]: This type of branching is essentially the same as what we use in the linear Quad element.

File: src/elements/ChrPlasmaLens.H (hunk 1 above)
Comment (kyrsjo) [reply src/elements/ChrPlasmaLens.H]: Done on my changes.

Should this really be applied to every single litteral in the code? E.g. also here:
https://github.com/BLAST-ImpactX/impactx/pull/1030/files#diff-3858396255f74cba0b31e856f293d63c9ee64f702310a4466d58a0c1c8cfe451R63 , in ConstF constructor argument list, which includes the code
```
amrex::ParticleReal dx = 0,
amrex::ParticleReal dy = 0,
amrex::ParticleReal rotation_degree = 0,
amrex::ParticleReal aperture_x = 0,
amrex::ParticleReal aperture_y = 0,
```

File: src/elements/ChrPlasmaLens.H (hunk 18 above)
Comment (cemitch99) [reply src/elements/ChrPlasmaLens.H:205]: A possible alternative to branching (using ternary operators) appears in the linear map of the CFbend element, line 262 [here](https://github.com/BLAST-ImpactX/impactx/blob/development/src/elements/CFbend.H).

File: src/elements/ChrPlasmaLens.H (hunk 1 above)
Comment (ax3l) [reply src/elements/ChrPlasmaLens.H]: For immediate assignments it does not matter that much, especially for zeros.

It is more problematic when passed into math functions and math terms, which causes doubly up and down-casting when we compile for single precision.

I just checked, and comparison to zero is actually no problem as well: https://godbolt.org/z/qYn8rfK8v

Sorry for the unnecessary comment!

File: src/elements/ConstF.H (hunk 8 above)
Comment (ax3l) [reply src/elements/ConstF.H]: Oh no! I thought it replaces it to four spaces each :)

Visually, it might depend what you see on what your editor was configured to show for tabs before conversion?

File: src/elements/ChrPlasmaLens.H (hunk 18 above)
Comment (ax3l) [reply src/elements/ChrPlasmaLens.H:205]: > Sure, I was assuming that the compiler would smooth it out anyway.

If we are lucky, it could find the duplication. But I think because we do not yet use fast-math, the subtle differences in signs will prevent it.

Also no problem to leave for now, we can refactor it later.

**Comment by ax3l** [comment]
Thanks a lot @kyrsjo !

> Is it OK that we keep the elements in 1 PR? I'll separate out the m_slice_ds uninitialized stuff to a separate PR.

I would really prefer to do one PR for each, because that way we can merge the quad right away and then decide for ConstF and plasma lens if and how much testing we add, too. Also makes the [release notes that we auto-generate](https://github.com/BLAST-ImpactX/impactx/releases) readable.

I already moved the Quad fix into #1038. Please merge in `development` here and it will vanish from the diff in this PR.

Please do the same for ConstF :pray:

**Review by cemitch99 — COMMENTED** [review]
I made some small comments, and will try to help pin down why the CI tests are currently failing.

File: src/elements/Quad.H (hunk 17 above)
Comment (ax3l) [reply src/elements/Quad.H:284]: Merged in #1038. Please rebase against `development` or merge the latest `development` :)

File: src/elements/ChrPlasmaLens.H [hunk 19]
```diff
@@ -232,7 +254,7 @@ namespace impactx::elements
             refpart.t = t - step*pt;
 
             // advance integrated path length
-            refpart.s = s + m_slice_ds;
+            refpart.s = s + slice_ds;
```
Comment (kyrsjo) [inline src/elements/ChrPlasmaLens.H:294]: Also move this to a separate PR?

File: src/elements/Quad.H (hunk 17 above)
Comment (kyrsjo) [reply src/elements/Quad.H:284]: Will do!

File: src/elements/ChrPlasmaLens.H (hunk 19 above)
Comment (ax3l) [reply src/elements/ChrPlasmaLens.H:294]: I would suggest to leave the plasma lens here and move the `ConstF` in a separate PR.

**Comment by kyrsjo** [comment]
OK, done. The changes to ConstF i put into PR #1041 and removed from this branch (just copied the contents of the `ConstF.h` file from the current development branch on top of my changes).

I also merged from upstream/development (before I split off the ConstF changes.)

Thanks a lot for the feedback - its really needed to get into the way of doing things in a new-to-me project!

File: src/elements/ChrPlasmaLens.H (hunk 19 above)
Comment (kyrsjo) [reply src/elements/ChrPlasmaLens.H:294]: Done.

**Comment by cemitch99** [comment]
@kyrsjo :  The problem is with the test `apochromatic_pl`.  If normalized focusing strength units are used (units = 0), then the test works.  The test also works if positrons are used instead of electrons.

The test inputs were impacted by PR[#935](https://github.com/BLAST-ImpactX/impactx/pull/935), and not yet updated.  The sign of the field did not matter previously, since it was assumed that k>0.  After this PR, the sign is important.  To use electrons, the sign of each focusing field (when unit=1) therefore needs to be changed (<0).

I can't push to your branch - can you modify the signs used for k in `input_apochromatic_pl.in` and `run_apochromatic_pl.py push` to this PR?  That should fix.

**Comment by kyrsjo** [comment]
Thanks for debugging! I got a bit confused for a while since the `.in` file seems to be ignored by ctest (e.g. `ctest --test-dir build -R apochromat_pl --output-on-failure`), but after also finding the Python run file, it worked.

**Comment by kyrsjo** [comment]
@cemitch99 weird that you cannot push to branch btw - ` Allow edits by maintainers ` is on. I'll just add you both to the fork repo.

Thanks again for the feedback! Sorry that I'm slow these days - it's holiday time.

File: src/elements/ChrPlasmaLens.H (hunk 18 above)
Comment (kyrsjo) [reply src/elements/ChrPlasmaLens.H:205]: OK, let's leave it for now then.

**Comment by kyrsjo** [comment]
I see there are some conflicts after the changes in PR #1002 - will fix in branch.

File: examples/apochromatic/input_apochromatic_pl.in [hunk 20]
```diff
@@ -35,43 +35,43 @@ dr2.ds = 2.0
 
 q1.type = plasma_lens_chromatic
 q1.ds = 0.331817852986604588
-q1.k = 996.147787384348956
+q1.k = -996.147787384348956
```
Comment (ax3l) [inline examples/apochromatic/input_apochromatic_pl.in:38]: Maybe I am just confused, byt why are all the lenses now set to defocusing (or are they)?

Do we still cover the positive k case in CI somewhere?

Comment (cemitch99) [reply examples/apochromatic/input_apochromatic_pl.in:38]: @ax3l Since `q1.units = 1`, the field is given in SI units.   Since these are electrons, the focusing field is negative.

**Comment by kyrsjo** [comment]
FYI I'm working on a test for a defocusing lens.

Is there a test for a lens that's off?

File: examples/apochromatic/input_apochromatic_pl.in (hunk 20 above)
Comment (ax3l) [reply examples/apochromatic/input_apochromatic_pl.in:38]: But the unit was not changed... was this a bug before the PR?

Comment (kyrsjo) [reply examples/apochromatic/input_apochromatic_pl.in:38]: We have `T_Real const omega = sqrt(abs(m_g)/delta1);` and `m_g = m_unit == 1 ? m_k / refpart.rigidity_Tm() : m_k;`.

In the old code version `omega` is always used with focusing matrix elements, whatever the sign of `m_g` or `k`. So the sign of `m_g` was effectively discarded, and it was forced to be a focusing element.

File: src/elements/ChrPlasmaLens.H [hunk 21]
```diff
@@ -265,7 +318,9 @@ namespace impactx::elements
         // see: compute_constants() to refresh
         amrex::ParticleReal m_slice_ds;  //! m_ds / nslice();
         amrex::ParticleReal m_beta;
+        amrex::ParticleReal m_gamma;
```
Comment (kyrsjo) [inline src/elements/ChrPlasmaLens.H:321]: Should these be `T_Real` as well? I'm not really sure about the difference vs `amrex::ParticleReal`.

This method gets called pre-tracking, and the variables get used in the tracking calculations.

File: examples/active_plasma_lens/analysis_APL_ChrPlasmaLens_defocusing.py [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+s_ref = beam_final.get_attribute("s_ref")
+gamma_ref = beam_final.get_attribute("gamma_ref")
+print(f"  sigx={sigx:e} sigy={sigy:e} sigt={sigt:e}")
+print(
+    f"  emittance_x={emittance_x:e} emittance_y={emittance_y:e} emittance_t={emittance_t:e}\n"
+    f"  s_ref={s_ref:e} gamma_ref={gamma_ref:e}"
+)
+
+atol = 0.0  # ignored
+# rtol = 2.2 * num_particles**-0.5  # from random sampling of a smooth distribution
```
Comment (kyrsjo) [inline examples/active_plasma_lens/analysis_APL_ChrPlasmaLens_defocusing.py:65]: Where does this 2.2 actually come from? For some reason, the sample sigma seems to always start a bit high, and with defocusing this is blown up to *just* over the limit.

I relaxed the constrain to let it pass, but I'm not sure if it's the right thing to do.

image

File: examples/active_plasma_lens/run_APL_ChrPlasmaLens.py [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+    # run simulation
+    sim.track_particles()
+
+    # clean shutdown
+    sim.finalize()
+
+
+def analytic_final_estimate(APL_g, rigidity_Tm, APL_length, beta_0, alpha_0):
+    ## Note: this seems to not work correctly for k != 0???
+    k = APL_g * rigidity_Tm
```
Comment (kyrsjo) [inline examples/active_plasma_lens/run_APL_ChrPlasmaLens.py]: This should be `/` not `*`

Comment (kyrsjo) [reply examples/active_plasma_lens/run_APL_ChrPlasmaLens.py]: Fixed.

**Comment by kyrsjo** [comment]
I think this is ready for review/merging now, @ax3l & @cemitch99 ?

There are two questions I wonder about (datatype and tolerance) + 1 question I've answered (why negative).

Note that I've made the tests so that I can fit the ConstF tests into the same framework :)

File: examples/active_plasma_lens/analysis_APL_ChrPlasmaLens_defocusing.py (hunk 22 above)
Comment (cemitch99) [reply examples/active_plasma_lens/analysis_APL_ChrPlasmaLens_defocusing.py:65]: The 2.2 factor is rather arbitrary.  We know the statistical fluctuations in the moments due to finite N should scale as sqrt(N), and this proportionality factor is chosen largely by trial-and-error.  (We start large, so that the test passes, and then decrease this value so that tests pass barely.)  Relaxing the constraint is just fine.

Comment (ax3l) [reply examples/active_plasma_lens/analysis_APL_ChrPlasmaLens_defocusing.py:65]: Exactly. Please feel free to slighly relax the constant from 2.2 to a higher number if need be.

File: src/elements/ChrPlasmaLens.H (hunk 21 above)
Comment (ax3l) [reply src/elements/ChrPlasmaLens.H:321]: No, because `m_gamma` and others calculated in `compute_constants` are independent of individual particles (they might at best depend on the reference particle).

We use `T_Real` only in the per-particle push to enable SIMD vectorization (pushing multiple particles at once). When we use `-DImpactX_SIMD=ON`, the types behind `T_Real` will do independent, point-wise math (roughly: with an overloaded `double[N]` or `float[N]`) instead of scalar math (`amrex::ParticleReal` aka `double`/`float`).

File: examples/active_plasma_lens/run_APL.py [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+    sigmap_0 = math.sqrt(emitg * gamma_0)
+    mu_0 = alpha_0 / math.sqrt(beta_0 * gamma_0)
+    print(
+        f"sigma_0 = {sigma_0} [m], beta_0 = {beta_0} [m], alpha_0 = {alpha_0}, sigmap_0 = {sigmap_0}"
+    )
+    print()
+
+    # Forward-propagate that through the focusing/defocusing lens
+    # from the beginning, ignoring energy spread
+    # [Doesn't give same result as ChrPlasmaLens and ChrQuad for some reason, even when sigpt_0 = 0]
```
Comment (cemitch99) [inline examples/active_plasma_lens/run_APL.py:75]: Does this comment still apply?  If so, this needs to be addressed.

File: examples/active_plasma_lens/analysis_APL_ChrPlasmaLens.py [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+    beta_x = sigx2 / emittance_x
+    alpha_x = -epstrms["position_x"]["momentum_x"] / emittance_x
+
+    sigy2 = epstrms["position_y"]["position_y"]
+    sigpy2 = epstrms["momentum_y"]["momentum_y"]
+    emittance_y = (sigy2 * sigpy2 - epstrms["position_y"]["momentum_y"] ** 2) ** 0.5
+    beta_y = sigy2 / emittance_y
+    alpha_y = -epstrms["position_y"]["momentum_y"] / emittance_y
+
+    return (beta_x, beta_y, alpha_x, alpha_y)
```
Comment (cemitch99) [inline examples/active_plasma_lens/analysis_APL_ChrPlasmaLens.py:53]: Note:  This works, but the values defined here are computed internally and stored as reduced_diagnostics.  See this example for an alternative approach:  https://impactx.readthedocs.io/en/latest/usage/examples/fodo_space_charge/README.html
It's fine to keep this for now.

**Review by cemitch99 — APPROVED** [review]

File: examples/active_plasma_lens/plot_APL_ChrPlasmaLens_analytical.py [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+# License: BSD-3-Clause-LBNL
+#
+# -*- coding: utf-8 -*-
+
+## Plots analytical Twiss parameters
+#  as a function of APL gradient [T/m]
+
+import matplotlib.pyplot as plt
+import numpy as np
+from run_APL_ChrPlasmaLens import analytic_final_estimate
```
Comment (cemitch99) [inline examples/active_plasma_lens/plot_APL_ChrPlasmaLens_analytical.py]:
```suggestion
from run_APL import analytic_final_estimate
```

Comment (cemitch99) [reply examples/active_plasma_lens/plot_APL_ChrPlasmaLens_analytical.py]: Just caught this - name change.  I just tried it.

File: docs/source/usage/examples.rst [hunk 27]
```diff
@@ -41,6 +41,7 @@ Single Particle Dynamics
    examples/reversibility/README.rst
    examples/charge_sign/README.rst
    examples/symplectic_integration/README.rst
+   examples/active_plasma_lens
```
Comment (ax3l) [inline docs/source/usage/examples.rst]:
```suggestion
   examples/active_plasma_lens/README.rst
```

Comment (cemitch99) [inline docs/source/usage/examples.rst]:
```suggestion
   examples/active_plasma_lens/README.rst
```

**Review by ax3l — APPROVED** [review]
Looks great, thank you for this! :rocket: :sparkles:

File: examples/active_plasma_lens/README.rst [hunk 28]
```diff
@@ -0,0 +1,45 @@
+.. _examples-active_plasma_lens:
+
+Active Plasma Lens
+==================
+
+These examples demonstrates the of the effect of an Active Plasma Lens (APL) on the beam.
```
Comment (cemitch99) [inline examples/active_plasma_lens/README.rst]:
```suggestion
These examples demonstrate the effect of an Active Plasma Lens (APL) on the beam.
```

File: examples/active_plasma_lens/README.rst [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+The lattice contains this element and nothing else.
+The length of the element is 20 mm, and it can be run in no-field, focusing, and defocusing mode.
+
+We use a 200 MeV electron beam with an initial normalized rms emittance of 10 um.
+The beam is set to have :math:`\alpha = 0` in the middle of the lens in the case of no field.
+The beam size in the middle of the lens is set to 10 µm for the no-field examples (in order to have a strongly parabolic :math:`\beta`-function within the lens), and 100 µm for the focusing and defocusing examples.
+A :math:`\sigma_{pt} = 10^{-3}` is also assumed.
+Before the simulations, this beam is back-propagate to the lens entry asuming no field.
```
Comment (cemitch99) [inline examples/active_plasma_lens/README.rst]:
```suggestion
Before the simulation, this beam is back-propagated to the lens entry assuming zero field.
```

File: examples/active_plasma_lens/plot_APL_ChrPlasmaLens_analytical.py (hunk 26 above)
Comment (kyrsjo) [reply examples/active_plasma_lens/plot_APL_ChrPlasmaLens_analytical.py]: Thank you!

File: examples/active_plasma_lens/run_APL.py (hunk 24 above)
Comment (kyrsjo) [reply examples/active_plasma_lens/run_APL.py:75]: Hi, this was fixed here:

https://github.com/BLAST-ImpactX/impactx/pull/1030/commits/bb670a1f2877cb999bc6f13a30728eb7e5f62c8c#diff-a95909e2ce0d4e807a3fa1dc18e09c9c822ec1dbb1275b60b056c775609e216eL125

and then I started using the analytical estimate here:
https://github.com/BLAST-ImpactX/impactx/pull/1030/commits/7f06ab254a786cb5e946063a241fcde4dd0723a0

I will cleanup when I do constF, PR #1041 - the tests should share quite a bit of code.

**Comment by kyrsjo** [comment]
Thank you @ax3l and @cemitch99 for merging! I will now proceed to ConstF. Also, PhD student @aellings will now be able to simulate diverging APLs in ImpactX :)

Maybe see you at Elba?

**Comment by ax3l** [comment]
@kyrsjo Awesome, thanks for the contribution!!

Yes, I am at EAAC until Thursday noon, let's catch up!
